onnx/shape_inference/implementation.h

```diff
@@ -241,6 +241,148 @@ struct InferenceContextImpl : public InferenceContext {
       graphAttributeInferencers_;
 };
 
+struct DataPropagationContextImpl : public DataPropagationContext {
+  DataPropagationContextImpl(
+      NodeProto& n,
+      const std::unordered_map<std::string, TypeProto*>& valueTypesByName,
+      const std::unordered_map<std::string, const TensorProto*>& inputDataByName,
+      std::unordered_map<std::string, TensorShapeProto>& generatedShapeData,
+      const GraphInferenceContext* graphInferenceContext = nullptr)
+      : generatedShapeData_{generatedShapeData}, graphInferenceContext_{graphInferenceContext} {
+
+    for (auto& attr : *n.mutable_attribute()) {
+      attributesByName_[attr.name()] = &attr;
+      if (attr.has_g()) {
+        // need a mutable GraphProto to run inferencing on this attribute
+        graphProtoAttributesByName_[attr.name()] = attr.mutable_g();
+      }
+    }
+
+    size_t input_idx = 0;
+    for (const auto& input : n.input()) {
+      inputIndexToNameMap_.insert({input_idx++, input});
+
+      auto valueTypesIter = valueTypesByName.find(input);
+      if (valueTypesIter != valueTypesByName.end()) {
+        allInputTypes_.push_back(valueTypesIter->second);
+      } else {
+        allInputTypes_.push_back(nullptr);
+      }
+
+      const auto inputDataIter = inputDataByName.find(input);
+      if (inputDataIter != inputDataByName.cend()) {
+        allInputData_.push_back(inputDataIter->second);
+      } else {
+        allInputData_.push_back(nullptr);
+      }
+    }
+
+    size_t output_idx = 0;
+    for (const auto& output : n.output()) {
+      outputIndexToNameMap_.insert({output_idx++, output});
+    }
+
+    allOutputTypes_.resize(n.output_size());
+  }
+
+  const AttributeProto* getAttribute(const std::string& name) const override {
+    auto iter = attributesByName_.find(name);
+    if (iter == attributesByName_.end()) {
+      return nullptr;
+    } else {
+      return iter->second;
+    }
+  }
+
+  size_t getNumInputs() const override {
+    return allInputTypes_.size();
+  }
+
+  const TypeProto* getInputType(size_t index) const override {
+    if (index >= allInputTypes_.size()) {
+      ONNX_THROW("Input " + ONNX_NAMESPACE::to_string(index) + " is out of bounds.");
+    }
+    return allInputTypes_[index];
+  }
+
+  size_t getNumOutputs() const override {
+    return allOutputTypes_.size();
+  }
+
+  const TypeProto* getOutputType(size_t index) const override {
+    if (index >= allOutputTypes_.size()) {
+      ONNX_THROW("Output " + ONNX_NAMESPACE::to_string(index) + " is out of bounds.");
+    }
+    return &allOutputTypes_[index];
+  }
+
+  // Convert integer vector into TensorShapeProto
+  template <typename INTEGER>
+  TensorShapeProto vectorToTensorShapeProto(const std::vector<INTEGER>& input_vals) const {
```

Comment:
nit: you can take in TensorShapeProto& tsp as an input param... this way you dont have to create and copy this obj